lib/schemes/refresh.js

```diff
@@ -0,0 +1,329 @@
+      }
+    }
+
+    // Only add client id to payload if enabled
+    if (dataClientIdProperty) {
+      endpoint.data[dataClientIdProperty] = this.$auth.getClientId(this.name)
+    }
+
+    // Only add grant type to payload if enabled
+    if (dataGrantTypeProperty !== false) {
+      endpoint.data[dataGrantTypeProperty] = grantType
+    }
+
+    // Try to fetch user and then set
+    return this.$auth.requestWith(
+      this.name,
+      endpoint,
+      this.options.endpoints.refresh
+    ).then(response => {
+      this._updateTokens(response)
+    }).catch(() => {
+      // TODO: Unhandled error
+      this._logoutLocally()
+    })
+  }
+
+  _scheduleTokenRefresh () {
+    // If auto refresh is disabled, bail
+    if (this.options.disableAutoRefresh) return
+
+    let intervalDuration = (this._getExpiration() - Date.now()) * 0.75
+    if (intervalDuration < 1000) {
+      // in case you misconfigured refreshing this will save your auth-server from a self-induced DDoS-Attack
+      intervalDuration = 1000
+    }
+
+    this.refreshInterval = setInterval(() => {
+      this._tokenRefresh()
+    }, intervalDuration)
+  }
+
+  _initializeErrorInterceptor () {
+    this.$auth.ctx.app.$axios.onError(error => {
+      // If error has config and response and error status is 401
+      if (error.config && error.response && error.response.status === 401) {
+        const token = this.$auth.getToken(this.name)
+        const headerToken = error.config.headers[this.options.tokenName]
+
+        // If current token is equal to header token
+        if (token && headerToken && headerToken === token) {
+          // Add the failed request config to failedRequestQueue
+          if (!error.config.__isRetryRequest) {
+            error.config.__isRetryRequest = true
+            this.failedRequestQueue.push(error.config)
+          }
+
+          // If is not refreshing the token
+          if (!this.isRefreshing) {
+            // Refresh it
+            return this._tokenRefresh().then(() => {
+              this._syncExpiration()
+              this._scheduleTokenRefresh()
+            })
+          } else {
+            // Or resolve Promise
+            return Promise.resolve()
+          }
+        } else if (token && headerToken && headerToken !== token && !error.config.__isRetryRequest) {
+          // Or just retry the request if token is not equal to header token
+          error.config.__isRetryRequest = true
+          error.config.headers[this.options.tokenName] = token
+          return this.$auth.ctx.app.$axios.request(error.config)
+        }
+      }
+
+      // Otherwise, reject Promise
+      return Promise.reject(error)
+    })
+  }
+
+  async mounted () {
+    if (this.options.tokenRequired) {
+      const token = this.$auth.syncToken(this.name)
+      this._setToken(token)
+      this._syncRefreshToken()
+
+      const expiration = this._syncExpiration()
+      if (Date.now() >= expiration) {
```

Comment:
We don't have to logout if only the token is expired. I think we should save the refresh token's expiration as well. I propose starting the mounted hook with calling `_tokenRefresh` (if we have tokens after `syncToken` and `syncRefreshToken`). `_tokenRefresh` could be responsible for checking the token expiration, and logging out the user (as it already does) if the refresh token is expired.

Reply:
I understand your point. I agree with you. But we could keep the option to logout the user if (access)token has expired, disabled by default. I know it's not common, but I need it in my use case.